eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentPacket.scala

```diff
@@ -213,9 +213,12 @@ object OutgoingPaymentPacket {
    *         - the trampoline onion to include in final payload of a normal onion
    */
   def buildTrampolineToLegacyPacket(invoice: PaymentRequest, hops: Seq[NodeHop], finalPayload: PaymentOnion.FinalPayload): (MilliSatoshi, CltvExpiry, Sphinx.PacketAndSecrets) = {
-    val (firstAmount, firstExpiry, payloads) = hops.drop(1).reverse.foldLeft((finalPayload.amount, finalPayload.expiry, Seq[PaymentOnion.PerHopPayload](finalPayload))) {
+    // NB: the final payload will never reach the recipient, since the next-to-last node in the trampoline route will convert that to a non-trampoline payment.
+    // We use the smallest final payload possible, otherwise we may overflow the trampoline onion size.
+    val dummyFinalPayload = PaymentOnion.createSinglePartPayload(finalPayload.amount, finalPayload.expiry, finalPayload.paymentSecret, None)
+    val (firstAmount, firstExpiry, payloads) = hops.drop(1).reverse.foldLeft((finalPayload.amount, finalPayload.expiry, Seq[PaymentOnion.PerHopPayload](dummyFinalPayload))) {
       case ((amount, expiry, payloads), hop) =>
-        // The next-to-last trampoline hop must include invoice data to indicate the conversion to a legacy payment.
+        // The next-to-last node in the trampoline route must receive invoice data to indicate the conversion to a non-trampoline payment.
```

Comment:
It would be good to have integration tests to check that the payment metadata works well with trampoline.

Reply:
I'll add a check.

Reply:
Actually it's not that simple, this data isn't exposed end-to-end (since we don't want to add it to the payments DB).
Do you have a suggestion on how we could test that without changing too much the internals?

I beefed up a bit the unit test coverage, but I agree it would be good to be able to test it end-to-end (but without modifying the business logic too much).

Reply:
The simplest option I see is to emit a new event, which contains the `paymentMetadata` we receive, but that seems wasteful to emit a new event for every payment received only for tests (but maybe this is ok?).

Reply:
We don't do anything with the metadata but it will need to be exposed somehow to be useful. But for now I guess the existing tests are enough.

Reply:
Yeah it would be nice to be able to test it...

I don't think we will expose it though: the way I'd like to use it is to "replace" what we store in the DB, so when we receive a `payment_metadata` we would decrypt it and extract the values inside, and if they are looking good, we would accept the payment, and otherwise reject it.

If we only do that it will stay "hidden" in our business logic internals. But we probably want applications to also be able to handle "db-less" payments, in that case we would probably have a mechanism for an external app to "register" and listen to a specific `payment_metadata`?

Reply:
> The simplest option I see is to emit a new event, which contains the paymentMetadata we receive, but that seems wasteful to emit a new event for every payment received only for tests (but maybe this is ok? I don't actually know how expensive that is).

It's not expensive, and even if it were, you can do it only in the tests by setting your own `paymentHandler`, like we already do in integration tests:
https://github.com/ACINQ/eclair/blob/27579a5786db083700bb668a93eb86583bdf9d74/eclair-core/src/test/scala/fr/acinq/eclair/integration/ChannelIntegrationSpec.scala#L358

Reply:
You're ignoring what happens outside of eclair, if the `payment_metadata` is correct, we should not just accept the payment and be done with it, we should also deliver the item that was bought to the buyer and for that we need the address that would be stored in the `payment_metadata`. Currently we expose the `payment_hash` so that the delivery system (or whatever comes after the payment) can look up the information it needs in its database, if we want to switch to not using a database, we should think about that delivery system too, not just the DB inside eclair.

Reply:
> You're ignoring what happens outside of eclair, if the payment_metadata is correct, we should not just accept the payment and be done with it, we should also deliver the item that was bought to the buyer and for that we need the address that would be stored in the payment_metadata. Currently we expose the payment_hash so that the delivery system (or whatever comes after the payment) can look up the information it needs in its database, if we want to switch to not using a database, we should think about that delivery system too, not just the DB inside eclair.

Yep, that's exactly what I'm saying, we will need to build such a system.
But since we don't have the protocol sketched out at all, we don't know yet how to expose this properly, so we have to wait (and we need to wait for senders to support this feature before we can hope to actually use it, so we have many months before that's close to ready).

Reply:
> you can do it only in the tests by setting your own paymentHandler, like we already do in integration tests:

Good idea, I'll try that!

Reply:
I went for the simplest version possible in https://github.com/ACINQ/eclair/pull/2063/commits/cb2aa019d62aed38c288786de59540b7df3fe580, where I just emit an event. We can rework this when we get closer to implementing features based on `payment_metadata`.